[FEAT] Waveform Component 추가

nakisara01 opened the pull request:
### 1. 🚀 어떤 종류의 PR인가요? (What type of PR is this?)
- [ ] **Bug**: 예상과 다르게 동작하는 문제 또는 버그를 수정합니다.
- [ ] **Chore**: 빌드, 설정, 라이브러리 업데이트 등 사용자에게 직접적인 영향이 없는 작업을 합니다.
- [ ] **Documentation**: 문서(README, Wiki, API 문서 등)를 추가하거나 수정합니다.
- [x] **Feature**: 새로운 기능을 추가합니다.
- [ ] **Refactor**: 버그 수정이나 기능 추가 없이 기존 코드를 개선합니다.
- [ ] **Style**: 코드 포맷팅, 세미콜론 누락 등 코드 스타일에 관련된 것을 수정합니다.
- [ ] **Test**: 테스트 코드를 추가하거나 수정합니다.


### 2. 📝 개요 (Description)
➡️ *가이드: PR의 목적을 한두 문장으로 요약합니다. '무엇을' & '왜' 수정했는지 리뷰어가 쉽게 이해할 수 있도록 작성하는 것이 핵심입니다.*
개발되어 있던 WaveformView를 적용했습니다.

### 3. 🔗 관련 이슈 (Related Issue)
➡️ *가이드: 관련된 이슈 번호를 `Closes #이슈번호` 형식으로 적어주세요. 이렇게 하면 PR이 머지될 때 해당 이슈가 자동으로 종료되어 편리합니다.*
Closes #13 


### 4. ✅ 작업 목록 (Checklist
➡️ *가이드: 리뷰어가 코드 변경 사항을 쉽게 따라갈 수 있도록, 실제 작업 내역을 구체적인 단위로 나누어 나열합니다.*
- WaveformExtrator: 음원 파일을 다운샘플링하여 진폭배열을 만들어주는 함수파일입니다.
- WaveformView: Waveform 시각화 및 이를 위한 수학적 계산 로직이 포함된 파일입니다.
- WaveformUseView: Waveform이 사용되는 형식을 보여주기 위해 제작한 뷰입니다. 추후에 Waveform을 ChordProgressView에 적용했거나 사용 양식을 확인하여 해당 파일이 필요없을 경우 삭제해도 무방합니다.

### 5. 📸 스크린샷 (Screenshots)
➡️ *가이드: UI 변경이 있다면, 수정 전(Before)과 후(After)의 모습을 스크린샷으로 첨부하여 리뷰어가 직관적으로 변화를 이해하도록 돕습니다.*
<img width="290" height="700" alt="스크린샷, 2025-10-22 오전 3 48 48" src="https://github.com/user-attachments/assets/27e24e32-bf1f-4521-9885-f0a6803dc497" />

### 6. ❗ 리뷰어에게 전달할 기타 사항 (Other information for reviewers)
➡️ *가이드: 리뷰어가 놓치지 않고 꼭 확인해야 할 핵심 코드 위치나, 특정 방식으로 테스트가 필요한 경우 친절하게 안내합니다. 리뷰 효율을 크게 높일 수 있습니다.*

nakisara01 committed:
[feat] Add Waveform Extractor, WaveformView

nakisara01 committed:
[feat] Add WaveformUseView

- Waveform 사용 예시뷰입니다. 추후 필요없을 경우 삭제해도 무방합니다.

nakisara01 committed:
[chore] Add sample audio & Waveform Background

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+struct WaveformExtractor {
+    
+    //MARK: 디코딩 파이프가 정상 작동하는 지 샘플 개수 세어서 확인하는 함수 -> 동작에는 크게 필요없지만 debug용으로 작동하는 함수
+    static func debugCountSamples(from url: URL, targetSampleRate: Double = 44_100) async throws -> Int {
```
결국 이 함수가 왜 필요한건가요? 파이프가 정상 작동하는 것과 샘플 개수를 세는 것과 무슨 연관이 있는지 PR 노트에 적어주면 이해하기 편할 것 같아요!

ehdrbdndns replied:
실제로 사용되지 않는 함수인 것 같은데욥! 언제 사용하는 걸 생각하고 만드신건가용?

nakisara01 replied:
추후에 디버깅용으로 사용하려고 만들었는데 지금 보니 필요 없는 함수인 것 같아서 삭제하겠습니다!

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+        ]
+        
+        let reader = try AVAssetReader(asset: asset)
+        let output = AVAssetReaderTrackOutput(track: track, outputSettings: outputSettings)
```
재사용되는 변수가 아니면, 따로 선언하지 않는게 가독성이 더 좋을 것 같아요

```suggestion
        let reader = try AVAssetReader(asset: asset)
        let output = AVAssetReaderTrackOutput(track: track, outputSettings: [
            AVFormatIDKey: kAudioFormatLinearPCM,
            AVLinearPCMIsFloatKey: true,
            AVLinearPCMIsBigEndianKey: false,
            AVLinearPCMBitDepthKey: 32,
            AVNumberOfChannelsKey: 1,
            AVSampleRateKey: targetSampleRate
        ])
```

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+        let totalSamplesEstimate = max(1, Int(durationSec * targetSampleRate))
+        let samplePerBin = max(1, totalSamplesEstimate / max(1, bins))
+        
+        // bin 단위로 accCount/accEnergy, accPeak를 모았다가 꽉 차면
```
주석 문구가 완성되지 않아서 이해하기 힘듭니다!

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+        var result = [Float](); result.reserveCapacity(bins)
+        var accCount = 0
+        var accEnergy: Float = 0
+        var accPeak: Float = 0
```
의도를 이해하기 어려운 변수명들인 것 같습니다

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+        var accEnergy: Float = 0
+        var accPeak: Float = 0
+        
+        func flushBin() {
```
flushBin은 그냥 bin 비우기라는 뜻만 가지고 있는 것 같아서 변경이 필요해보여요.

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+        let maxV = max(result.max() ?? 0, 1e-6)
+        let normalized = result.map { min(1, max(0, $0 / maxV)) }
+        
+        return normalized
```
보통 result 변수 명을 사용했을 경우에 result가 반환되어야 하지 않나 싶습니다!

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+            throw WaveformError.readerFailed(reader.error?.localizedDescription ?? "Unknown error")
+        }
+        
+        let maxV = max(result.max() ?? 0, 1e-6)
```
1e-6이 의미하는 바가 뭔지, 의도가 무엇인지 주석처리해주면 좋을 것 같아요

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+                    while i < count {
+                        let s = fptr[i]
+                        let a = abs(s)
+                        accPeak = max(accPeak, a)
```
p, s, a이런 변수 명들은 삼가주세요! 다른 사람들이 코드리뷰할 때 이해하기 어렵고 시간이 오래 걸리는 요소들인 것 같습니다

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformView.swift:
```diff
@@ -0,0 +1,75 @@
+                let maxBarHeight: CGFloat = 19
+
+                for i in 0..<amps.count {
+                    var v = CGFloat(amps[i])
+                    if v < amplitudeFloor { continue }
+                    v = (v - amplitudeFloor) / (1 - amplitudeFloor)
+
+                    let h = minBarHeight + (maxBarHeight - minBarHeight) * v
+                    let x = CGFloat(i) * step
+                    let y = (size.height - h) / 2
+                    let path = Path(roundedRect: CGRect(x: x, y: y, width: barWidth, height: h),
+                                    cornerRadius: barWidth / 2)
+
+                    // 색 결정: highlight > filled > 기본
+                    let isHL = mappedHL?.contains(i) ?? false
+                    let isFilled = i <= filledIndex
+                    let color: Color = {
+                        if isHL { return .green }
+                        if isFilled { return .green }
+                        return .gray.opacity(0.35)
+                    }()
```
Bar 컴포넌트를 하나 만드는게 좋을 것 같습니다.

h, x, y, v 등의 변수 명은 피해주세요. barHeight 이런식으로 정해주시면 좋을 것같아요.

isHL도 isHighlight가 좋을 것 같습니다. isPy라고 안하고 isPlaying이라고 하는 것처럼!

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformUseView.swift:
```diff
@@ -0,0 +1,167 @@
+    @State private var player: AVAudioPlayer?
+    @State private var timer: Timer?
+
+    private let bins = 150
```
bin의 값이 150인 이유에 대한 근거도 같이 남겨주면 좋을 것 같네요. PR Note라든지 주석으로. 만약 150이외에 설정할 값들이 없다면 그냥 전역상수로 두는 것이 좋을 듯 합니다.

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformUseView.swift:
```diff
@@ -0,0 +1,167 @@
+            player?.prepareToPlay()
+            player?.play()
+        } catch {
+            print("재생 실패:", error)
+            return
```
더 명확한 에러 처리가 필요해보입니다. 잘 생각이 안나면 차라리 throw로 우선 던지기...? 디자이너한테 system에러 날 시 디자인 만들어달라고해야겠네요.

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformView.swift:
```diff
@@ -0,0 +1,75 @@
+
+    var body: some View {
+        GeometryReader { geo in
+            let barWidth: CGFloat = 5.0
+            let barSpacing: CGFloat = 5.0
+            let step = barWidth + barSpacing
+            let barCount = max(1, Int(geo.size.width / step))
+            let amps = resample(amplitudes, to: barCount)
+            let amplitudeFloor: CGFloat = 0.0
+
```
뭔가.. 의도를 파악하기 어려워서, 함수로 묶어서 의도를 설명해주시면 좋을 것 같아요. init()에 넣는다던지..등등

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformUseView.swift:
```diff
@@ -0,0 +1,167 @@
+        }
+    }
+
+    struct SampleLoadButton: View {
+        @Binding var amps: [Float]
+        let bins: Int
+
+        var body: some View {
+            HStack {
+                Button("Load sample.m4a") {
+                    Task {
+                        do {
+                            guard let url = Bundle.main.url(forResource: "sample", withExtension: "m4a") else { return }
+                            amps = try await WaveformExtractor.extractAmplitudes(
+                                from: url, bins: bins, mode: .rms, targetSampleRate: 44_100
+                            )
+                        } catch { print(error) }
+                    }
+                }
+                .buttonStyle(.borderedProminent)
+            }
+        }
+    }
```
음성 파일을 진폭 배열로 변환하는 작업을 WavefromView 내부에서 처리하는게 좋을 것 같습니다.

ehdrbdndns commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+    }
+    
+    //MARK: 진폭 배열 추출 함수 (녹음본 다운샘플링)
+    static func extractAmplitudes(from url: URL, bins: Int, mode: DownsampleMode = .rms, targetSampleRate: Double = 44_100) async throws -> [Float] {
```
지금 디자인 된 모습을 보면 하나의 음성 파일에, 마디별로 파형이 나와야합니다.

그래서 extractAmplitudes에 시작시간, 끝 시간을 받아서 진폭 배열을 반환시켜야할 것 같아요.

ehdrbdndns replied:
bins, mode, targetSampleRate의 값에 변동을 줄 일이 없다면 그냥 전역상수 처리하는게 더 좋을 것 같네요!

ehdrbdndns reviewed (commented):
고생하셨습니다!

전반적으로 코멘트 남겼습니다! 

오늘도 화이팅!! 👏

Segyun commented on GMG/GMG/Resources/Assets.xcassets/WaveformBackground.imageset/Group 87.pdf:
파일명도 이미지 이름이랑 같이 통일하는게 좋을 것 같아요!

Segyun commented on GMG/GMG/UI/Components/Waveform/WaveformExtractor.swift:
```diff
@@ -0,0 +1,164 @@
+            var dataPointer: UnsafeMutablePointer<Int8>?
+            CMBlockBufferGetDataPointer(blockBuffer, atOffset: 0, lengthAtOffsetOut: nil, totalLengthOut: &length, dataPointerOut: &dataPointer)
+            
+            if let _ = dataPointer {
```
nil 검사만 하는 거라면 nil 비교 연산으로 수정하는 것이 좋을 것 같아요.

```suggestion
            if dataPointer != nil {
```

Segyun commented on GMG/GMG/UI/Components/Waveform/WaveformUseView.swift:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    struct PlaySection: View {
+        let isEnabled: Bool
```
상위 뷰에서 `.disabled` 모디파이어 사용하면 isEnabled 사용하지 않아도 됩니다!

```Swift
            PlaySection {
                playFromStart()
            }
            .disabled(amps.isEmpty)
```

```Swift
    struct PlaySection: View {
        let action: () -> Void

        init(action: @escaping () -> Void) {
            self.action = action
        }

        var body: some View {
            Button("Play", action: action)
        }
    }
```

혹은 값 자체가 필요하다면 `@Environment(\.isEnabled) private var isEnabled`로 사용할 수도 있어요!

Segyun commented on GMG/GMG/UI/Components/Waveform/WaveformUseView.swift:
```diff
@@ -0,0 +1,167 @@
+                    .resizable()
+                    .scaledToFit()
+                    .background(Color.white)
+                    .clipShape(RoundedCornerRectangle(radius: 20, corners: [.bottomLeft, .bottomRight]))
```
내장 Shape로도 구현할 수 있어요!

```suggestion
                    .clipShape(UnevenRoundedRectangle(bottomLeadingRadius: 20, bottomTrailingRadius: 20))
```

Segyun commented on GMG/GMG/UI/Components/Waveform/WaveformView.swift:
```diff
@@ -0,0 +1,75 @@
+                return min(max(idx, 0), max(0, amps.count - 1))
+            }()
+
+            Canvas { ctx, size in
```
기본 Shape View로도 구현 가능할 것 같은데 Canvas로 구현한 이유가 있을까요?

nakisara01 committed:
[fix] delete debugging function

nakisara01 committed:
[fix] improve variable efficiency

- 재사용되지 않는 변수에 대해 따로 선언하여 사용한 부분을 변경했습니다

nakisara01 committed:
[fix] flushBin name modify

- 각각의 bin에 대한 대표값을 지정하는 함수의 이름이 flushBin인 것이 적절치 않아 commitBin으로 변경하여 가독성이 좋도록 수정했습니다

nakisara01 committed:
[fix] modify unobvious variable name

- commitBin은 대표값 확정 정도가 아니라 값 산출 후 방출한다는 의미의 emitBin으로 변경
- 이외에 accPeak,accCount 등의 이해하기 어려운 함수 직관적인 변수 네임으로 변경

nakisara01 committed:
[fix] fix return value name

- return Value가 result가 아니라 normalized여서 혼동되는 점을 감안해 result를 binAmplitudes, return 값을 waveform으로 지정해 쉽게 이해할 수 있도록 조정했습니다.

nakisara01 committed:
[fix] explain about 1e-6

- 1e-6을 사용하는 이유에 대해 주석 추가했습니다.

nakisara01 committed:
[fix] fix short variable name

- p, s, a 같은 변수를 pointer, sample amplitude와 같이 이해하기 쉽도록 수정했습니다.

nakisara01 committed:
[fix] delete waveformBackground

ehdrbdndns commented:
feat/#40에서 구현 완료
